File: ios/PPSSignatureView.m
#import "PPSSignatureView.h"

#define STROKE_WIDTH_MIN 2.0
#define STROKE_WIDTH_MAX 7.0
#define STROKE_WIDTH_SMOOTHING 0.5

#define VELOCITY_CLAMP_MIN 20
#define VELOCITY_CLAMP_MAX 5000

#define QUADRATIC_DISTANCE_TOLERANCE 3.0

static float clamp(float min, float max, float value) { return fmaxf(min, fminf(max, value)); }

@interface PPSSignatureView () {
    CGContextRef _bitmapContext;
    CGFloat _penThickness;
    CGFloat _previousThickness;

    CGPoint _previousPoint;
    CGPoint _previousMidPoint;

    UIColor *_bgColor;
    UIColor *_penColor;
}

@end

@implementation PPSSignatureView

#pragma mark - Lifecycle

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self commonInit];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        [self commonInit];
    }
    return self;
}

- (void)commonInit {
    _bgColor = [UIColor whiteColor];
    _penColor = [UIColor blackColor];
    _penThickness = (STROKE_WIDTH_MIN + STROKE_WIDTH_MAX) / 2.0;
    _previousThickness = _penThickness;
    _previousPoint = CGPointMake(-100, -100);
    _previousMidPoint = CGPointMake(-100, -100);

    self.backgroundColor = _bgColor;
    self.opaque = NO;
    self.hasSignature = NO;

    UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(pan:)];
    pan.maximumNumberOfTouches = 1;
    pan.minimumNumberOfTouches = 1;
    pan.cancelsTouchesInView = YES;
    [self addGestureRecognizer:pan];

    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tap:)];
    tap.cancelsTouchesInView = YES;
    [self addGestureRecognizer:tap];

    UILongPressGestureRecognizer *longer = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(longPress:)];
    longer.cancelsTouchesInView = YES;
    [self addGestureRecognizer:longer];
}

- (void)dealloc {
    [self releaseBitmapContext];
}

- (void)releaseBitmapContext {
    if (_bitmapContext) {
        CGContextRelease(_bitmapContext);
        _bitmapContext = NULL;
    }
}

- (void)layoutSubviews {
    [super layoutSubviews];

    if (self.bounds.size.width <= 0 || self.bounds.size.height <= 0) {
        return;
    }

    // Recreate bitmap context when bounds change
    CGImageRef oldImage = NULL;
    if (_bitmapContext) {
        oldImage = CGBitmapContextCreateImage(_bitmapContext);
        [self releaseBitmapContext];
    }

    [self ensureBitmapContext];

    // Redraw old content into new context if we had one
    if (oldImage) {
        CGContextDrawImage(_bitmapContext, self.bounds, oldImage);
        CGImageRelease(oldImage);
    }
}

#pragma mark - Bitmap Context

- (void)ensureBitmapContext {
    if (_bitmapContext) return;

    CGFloat scale = [UIScreen mainScreen].scale;
    NSInteger width = (NSInteger)(self.bounds.size.width * scale);
    NSInteger height = (NSInteger)(self.bounds.size.height * scale);

    if (width <= 0 || height <= 0) return;

    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    _bitmapContext = CGBitmapContextCreate(
        NULL,
        width,
        height,
        8,
        width * 4,
        colorSpace,
        kCGImageAlphaPremultipliedLast
    );
    CGColorSpaceRelease(colorSpace);

    if (!_bitmapContext) return;

    // Scale for retina
    CGContextScaleCTM(_bitmapContext, scale, scale);

    // Flip coordinate system (CGContext is bottom-up, UIKit is top-down)
    CGContextTranslateCTM(_bitmapContext, 0, self.bounds.size.height);
    CGContextScaleCTM(_bitmapContext, 1.0, -1.0);

    // Fill with background color
    [self fillBitmapWithBackgroundColor];
}

- (void)fillBitmapWithBackgroundColor {
    if (!_bitmapContext) return;

    CGContextSaveGState(_bitmapContext);
    CGContextSetFillColorWithColor(_bitmapContext, _bgColor.CGColor);
    CGContextFillRect(_bitmapContext, self.bounds);
    CGContextRestoreGState(_bitmapContext);
}

#pragma mark - Drawing

- (void)drawRect:(CGRect)rect {
    if (!_bitmapContext) return;

    CGImageRef image = CGBitmapContextCreateImage(_bitmapContext);
    if (image) {
        CGContextRef ctx = UIGraphicsGetCurrentContext();
        if (ctx) {
            CGContextDrawImage(ctx, self.bounds, image);
        }
        CGImageRelease(image);
    }
}

#pragma mark - Gesture Recognizers

- (void)tap:(UITapGestureRecognizer *)t {
    if (t.state != UIGestureRecognizerStateRecognized) return;

    CGPoint l = [t locationInView:self];

    [self ensureBitmapContext];
    if (!_bitmapContext) return;

    CGFloat radius = _penThickness > 0 ? _penThickness : (STROKE_WIDTH_MIN + STROKE_WIDTH_MAX) / 2.0;

    CGContextSaveGState(_bitmapContext);
    CGContextSetFillColorWithColor(_bitmapContext, _penColor.CGColor);
    CGRect dotRect = CGRectMake(l.x - radius / 2.0, l.y - radius / 2.0, radius, radius);
    CGContextFillEllipseInRect(_bitmapContext, dotRect);
    CGContextRestoreGState(_bitmapContext);

    self.hasSignature = YES;
    [self setNeedsDisplay];
}

- (void)longPress:(UILongPressGestureRecognizer *)lp {
    if (lp.state == UIGestureRecognizerStateBegan) {
        [self erase];
    }
}

- (void)pan:(UIPanGestureRecognizer *)p {
    CGPoint v = [p velocityInView:self];
    CGPoint l = [p locationInView:self];

    CGFloat distance = 0.0;
    if (_previousPoint.x > -50) {
        distance = sqrtf((l.x - _previousPoint.x) * (l.x - _previousPoint.x) +
                         (l.y - _previousPoint.y) * (l.y - _previousPoint.y));
    }

    CGFloat velocityMagnitude = sqrtf(v.x * v.x + v.y * v.y);
    CGFloat clampedVelocityMagnitude = clamp(VELOCITY_CLAMP_MIN, VELOCITY_CLAMP_MAX, velocityMagnitude);
    CGFloat normalizedVelocity = (clampedVelocityMagnitude - VELOCITY_CLAMP_MIN) / (VELOCITY_CLAMP_MAX - VELOCITY_CLAMP_MIN);

    CGFloat lowPassFilterAlpha = STROKE_WIDTH_SMOOTHING;
    CGFloat newThickness = (STROKE_WIDTH_MAX - STROKE_WIDTH_MIN) * (1 - normalizedVelocity) + STROKE_WIDTH_MIN;
    _penThickness = _penThickness * lowPassFilterAlpha + newThickness * (1 - lowPassFilterAlpha);

    [self ensureBitmapContext];
    if (!_bitmapContext) return;

    if ([p state] == UIGestureRecognizerStateBegan) {
        _previousPoint = l;
        _previousMidPoint = l;
        _previousThickness = _penThickness;

        self.hasSignature = YES;
        if (self.onDraggedBlock) {
            self.onDraggedBlock();
        }

    } else if ([p state] == UIGestureRecognizerStateChanged) {
        CGPoint mid = CGPointMake((l.x + _previousPoint.x) / 2.0, (l.y + _previousPoint.y) / 2.0);

        if (distance > QUADRATIC_DISTANCE_TOLERANCE) {
            int segments = (int)(distance / 1.5);

            CGFloat startPenThickness = _previousThickness;
            CGFloat endPenThickness = _penThickness;
            _previousThickness = _penThickness;

            for (int i = 0; i < segments; i++) {
                CGFloat t = (CGFloat)i / (CGFloat)segments;
                CGFloat thickness = startPenThickness + ((endPenThickness - startPenThickness) / segments) * i;

                // Quadratic Bezier interpolation
                CGFloat a = pow((1.0 - t), 2.0);
                CGFloat b = 2.0 * t * (1.0 - t);
                CGFloat c = pow(t, 2.0);

                CGPoint quadPoint = CGPointMake(
                    a * _previousMidPoint.x + b * _previousPoint.x + c * mid.x,
                    a * _previousMidPoint.y + b * _previousPoint.y + c * mid.y
                );

                [self strokeSegmentFrom:_previousPoint to:quadPoint thickness:thickness];
            }
        } else if (distance > 1.0) {
            [self strokeSegmentFrom:_previousPoint to:l thickness:_penThickness];
            _previousThickness = _penThickness;
        }

        _previousPoint = l;
        _previousMidPoint = mid;

    } else if (p.state == UIGestureRecognizerStateEnded || p.state == UIGestureRecognizerStateCancelled) {
        // Draw final point
        [self strokeSegmentFrom:_previousPoint to:l thickness:_penThickness];
    }

    [self setNeedsDisplay];
}

- (void)strokeSegmentFrom:(CGPoint)from to:(CGPoint)to thickness:(CGFloat)thickness {
    if (!_bitmapContext) return;

    CGContextSaveGState(_bitmapContext);

    CGContextSetStrokeColorWithColor(_bitmapContext, _penColor.CGColor);
    CGContextSetLineWidth(_bitmapContext, thickness);
    CGContextSetLineCap(_bitmapContext, kCGLineCapRound);
    CGContextSetLineJoin(_bitmapContext, kCGLineJoinRound);

    CGContextMoveToPoint(_bitmapContext, from.x, from.y);
    CGContextAddLineToPoint(_bitmapContext, to.x, to.y);
    CGContextStrokePath(_bitmapContext);

    CGContextRestoreGState(_bitmapContext);
}

#pragma mark - Erase

- (void)erase {
    self.hasSignature = NO;
    _previousPoint = CGPointMake(-100, -100);
    _previousMidPoint = CGPointMake(-100, -100);
    _penThickness = (STROKE_WIDTH_MIN + STROKE_WIDTH_MAX) / 2.0;
    _previousThickness = _penThickness;

    if (_bitmapContext) {
        [self fillBitmapWithBackgroundColor];
    }

    [self setNeedsDisplay];
}

#pragma mark - Color Properties

- (void)setStrokeColor:(UIColor *)strokeColor {
    _strokeColor = strokeColor;
    _penColor = strokeColor ?: [UIColor blackColor];
}

- (void)setBackgroundColor:(UIColor *)backgroundColor {
    [super setBackgroundColor:backgroundColor];
    _bgColor = backgroundColor ?: [UIColor whiteColor];

    if (_bitmapContext && !self.hasSignature) {
        [self fillBitmapWithBackgroundColor];
        [self setNeedsDisplay];
    }
}

#pragma mark - Image Export

- (UIImage *)snapshot {
    if (!_bitmapContext) return nil;

    CGImageRef cgImage = CGBitmapContextCreateImage(_bitmapContext);
    if (!cgImage) return nil;

    UIImage *image = [UIImage imageWithCGImage:cgImage scale:[UIScreen mainScreen].scale orientation:UIImageOrientationUp];
    CGImageRelease(cgImage);
    return image;
}

- (UIImage *)rotateImage:(UIImage *)sourceImage clockwise:(BOOL)clockwise {
    CGSize size = sourceImage.size;
    UIGraphicsBeginImageContext(CGSizeMake(size.height, size.width));
    [[UIImage imageWithCGImage:[sourceImage CGImage]
                         scale:1.0
                   orientation:clockwise ? UIImageOrientationRight : UIImageOrientationLeft]
     drawInRect:CGRectMake(0, 0, size.height, size.width)];

    UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    return newImage;
}

- (UIImage *)reduceImage:(UIImage *)image toSize:(CGSize)newSize {
    CGSize scaledSize = newSize;
    float scaleFactor = 1.0;

    if (image.size.width > image.size.height) {
        scaleFactor = image.size.width / image.size.height;
        scaledSize.width = newSize.width;
        scaledSize.height = newSize.height / scaleFactor;
    } else {
        scaleFactor = image.size.height / image.size.width;
        scaledSize.height = newSize.height;
        scaledSize.width = newSize.width / scaleFactor;
    }

    UIGraphicsBeginImageContext(scaledSize);
    CGRect scaledImageRect = CGRectMake(0.0, 0.0, scaledSize.width, scaledSize.height);
    [image drawInRect:scaledImageRect];

    UIImage *scaledImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    return scaledImage;
}

- (UIImage *)signatureImage {
    return [self signatureImage:NO withSquare:NO];
}

- (UIImage *)signatureImage:(BOOL)rotatedImage {
    return [self signatureImage:rotatedImage withSquare:NO];
}

- (UIImage *)signatureImage:(BOOL)rotatedImage withSquare:(BOOL)square {
    if (!self.hasSignature)
        return nil;

    UIImage *signatureImg;
    UIImage *snapshotImg = [self snapshot];
    [self erase];

    if (square) {
        signatureImg = [self reduceImage:snapshotImg toSize:CGSizeMake(400.0f, 400.0f)];
    } else {
        signatureImg = snapshotImg;
    }

    if (rotatedImage) {
        signatureImg = [self rotateImage:signatureImg clockwise:NO];
    }

    return signatureImg;
}

@end

File: ios/PPSSignatureView.swift
import UIKit

private let strokeWidthMin: CGFloat = 2.0
private let strokeWidthMax: CGFloat = 7.0
private let strokeWidthSmoothing: CGFloat = 0.5

private let velocityClampMin: CGFloat = 20
private let velocityClampMax: CGFloat = 5000

private let quadraticDistanceTolerance: CGFloat = 3.0

@objcMembers
class PPSSignatureView: UIView {

    var strokeColor: UIColor? {
        didSet {
            penColor = strokeColor ?? .black
        }
    }

    var hasSignature: Bool = false

    var onDraggedBlock: (() -> Void)?

    // MARK: - Private state

    private var bitmapContext: CGContext?
    private var penThickness: CGFloat = (strokeWidthMin + strokeWidthMax) / 2.0
    private var previousThickness: CGFloat = (strokeWidthMin + strokeWidthMax) / 2.0

    private var previousPoint = CGPoint(x: -100, y: -100)
    private var previousMidPoint = CGPoint(x: -100, y: -100)

    private var bgColor: UIColor = .white
    private var penColor: UIColor = .black

    // MARK: - Lifecycle

    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        commonInit()
    }

    private func commonInit() {
        bgColor = .white
        penColor = .black
        penThickness = (strokeWidthMin + strokeWidthMax) / 2.0
        previousThickness = penThickness
        previousPoint = CGPoint(x: -100, y: -100)
        previousMidPoint = CGPoint(x: -100, y: -100)

        backgroundColor = bgColor
        isOpaque = false
        hasSignature = false

        let pan = UIPanGestureRecognizer(target: self, action: #selector(handlePan(_:)))
        pan.maximumNumberOfTouches = 1
        pan.minimumNumberOfTouches = 1
        pan.cancelsTouchesInView = true
        addGestureRecognizer(pan)

        let tap = UITapGestureRecognizer(target: self, action: #selector(handleTap(_:)))
        tap.cancelsTouchesInView = true
        addGestureRecognizer(tap)

        let longPress = UILongPressGestureRecognizer(target: self, action: #selector(handleLongPress(_:)))
        longPress.cancelsTouchesInView = true
        addGestureRecognizer(longPress)
    }

    deinit {
        bitmapContext = nil
    }

    override func layoutSubviews() {
        super.layoutSubviews()

        guard bounds.size.width > 0, bounds.size.height > 0 else { return }

        var oldImage: CGImage?
        if let ctx = bitmapContext {
            oldImage = ctx.makeImage()
            bitmapContext = nil
        }

        ensureBitmapContext()

        if let oldImage = oldImage, let ctx = bitmapContext {
            ctx.draw(oldImage, in: bounds)
        }
    }

    // MARK: - Bitmap Context

    private func ensureBitmapContext() {
        guard bitmapContext == nil else { return }

        let scale = UIScreen.main.scale
        let width = Int(bounds.size.width * scale)
        let height = Int(bounds.size.height * scale)

        guard width > 0, height > 0 else { return }

        let colorSpace = CGColorSpaceCreateDeviceRGB()
        guard let ctx = CGContext(
            data: nil,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: width * 4,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return }

        bitmapContext = ctx

        ctx.scaleBy(x: scale, y: scale)
        ctx.translateBy(x: 0, y: bounds.size.height)
        ctx.scaleBy(x: 1.0, y: -1.0)

        fillBitmapWithBackgroundColor()
    }

    private func fillBitmapWithBackgroundColor() {
        guard let ctx = bitmapContext else { return }
        ctx.saveGState()
        ctx.setFillColor(bgColor.cgColor)
        ctx.fill(bounds)
        ctx.restoreGState()
    }

    // MARK: - Drawing

    override func draw(_ rect: CGRect) {
        guard let bitmapCtx = bitmapContext,
              let image = bitmapCtx.makeImage(),
              let ctx = UIGraphicsGetCurrentContext() else { return }
        ctx.draw(image, in: bounds)
    }

    // MARK: - Gesture Recognizers

    @objc private func handleTap(_ t: UITapGestureRecognizer) {
        guard t.state == .recognized else { return }

        let l = t.location(in: self)

        ensureBitmapContext()
        guard let ctx = bitmapContext else { return }

        let radius = penThickness > 0 ? penThickness : (strokeWidthMin + strokeWidthMax) / 2.0

        ctx.saveGState()
        ctx.setFillColor(penColor.cgColor)
        let dotRect = CGRect(x: l.x - radius / 2.0, y: l.y - radius / 2.0,
                             width: radius, height: radius)
        ctx.fillEllipse(in: dotRect)
        ctx.restoreGState()

        hasSignature = true
        setNeedsDisplay()
    }

    @objc private func handleLongPress(_ lp: UILongPressGestureRecognizer) {
        if lp.state == .began {
            erase()
        }
    }

    @objc private func handlePan(_ p: UIPanGestureRecognizer) {
        let v = p.velocity(in: self)
        let l = p.location(in: self)

        var distance: CGFloat = 0.0
        if previousPoint.x > -50 {
            distance = sqrt((l.x - previousPoint.x) * (l.x - previousPoint.x) +
                            (l.y - previousPoint.y) * (l.y - previousPoint.y))
        }

        let velocityMagnitude = sqrt(v.x * v.x + v.y * v.y)
        let clampedVelocity = min(velocityClampMax, max(velocityClampMin, velocityMagnitude))
        let normalizedVelocity = (clampedVelocity - velocityClampMin) / (velocityClampMax - velocityClampMin)

        let newThickness = (strokeWidthMax - strokeWidthMin) * (1 - normalizedVelocity) + strokeWidthMin
        penThickness = penThickness * strokeWidthSmoothing + newThickness * (1 - strokeWidthSmoothing)

        ensureBitmapContext()
        guard bitmapContext != nil else { return }

        switch p.state {
        case .began:
            previousPoint = l
            previousMidPoint = l
            previousThickness = penThickness

            hasSignature = true
            onDraggedBlock?()

        case .changed:
            let mid = CGPoint(x: (l.x + previousPoint.x) / 2.0,
                              y: (l.y + previousPoint.y) / 2.0)

            if distance > quadraticDistanceTolerance {
                let segments = Int(distance / 1.5)
                let startPenThickness = previousThickness
                let endPenThickness = penThickness
                previousThickness = penThickness

                for i in 0..<segments {
                    let t = CGFloat(i) / CGFloat(segments)
                    let thickness = startPenThickness +
                        ((endPenThickness - startPenThickness) / CGFloat(segments)) * CGFloat(i)

                    let a = pow(1.0 - t, 2.0)
                    let b = 2.0 * t * (1.0 - t)
                    let c = pow(t, 2.0)

                    let quadPoint = CGPoint(
                        x: a * previousMidPoint.x + b * previousPoint.x + c * mid.x,
                        y: a * previousMidPoint.y + b * previousPoint.y + c * mid.y
                    )

                    strokeSegment(from: previousPoint, to: quadPoint, thickness: thickness)
                }
            } else if distance > 1.0 {
                strokeSegment(from: previousPoint, to: l, thickness: penThickness)
                previousThickness = penThickness
            }

            previousPoint = l
            previousMidPoint = mid

        case .ended, .cancelled:
            strokeSegment(from: previousPoint, to: l, thickness: penThickness)

        default:
            break
        }

        setNeedsDisplay()
    }

    private func strokeSegment(from: CGPoint, to: CGPoint, thickness: CGFloat) {
        guard let ctx = bitmapContext else { return }

        ctx.saveGState()
        ctx.setStrokeColor(penColor.cgColor)
        ctx.setLineWidth(thickness)
        ctx.setLineCap(.round)
        ctx.setLineJoin(.round)
        ctx.move(to: from)
        ctx.addLine(to: to)
        ctx.strokePath()
        ctx.restoreGState()
    }

    // MARK: - Erase

    func erase() {
        hasSignature = false
        previousPoint = CGPoint(x: -100, y: -100)
        previousMidPoint = CGPoint(x: -100, y: -100)
        penThickness = (strokeWidthMin + strokeWidthMax) / 2.0
        previousThickness = penThickness

        if bitmapContext != nil {
            fillBitmapWithBackgroundColor()
        }

        setNeedsDisplay()
    }

    // MARK: - Color Properties

    override var backgroundColor: UIColor? {
        didSet {
            bgColor = backgroundColor ?? .white
            if bitmapContext != nil && !hasSignature {
                fillBitmapWithBackgroundColor()
                setNeedsDisplay()
            }
        }
    }

    // MARK: - Image Export

    private func snapshot() -> UIImage? {
        guard let ctx = bitmapContext, let cgImage = ctx.makeImage() else { return nil }
        return UIImage(cgImage: cgImage, scale: UIScreen.main.scale, orientation: .up)
    }

    private func rotateImage(_ sourceImage: UIImage, clockwise: Bool) -> UIImage {
        guard let cgImage = sourceImage.cgImage else { return sourceImage }
        let size = sourceImage.size
        let renderer = UIGraphicsImageRenderer(size: CGSize(width: size.height, height: size.width))
        return renderer.image { _ in
            UIImage(
                cgImage: cgImage,
                scale: 1.0,
                orientation: clockwise ? .right : .left
            ).draw(in: CGRect(x: 0, y: 0, width: size.height, height: size.width))
        }
    }

    private func reduceImage(_ image: UIImage, to newSize: CGSize) -> UIImage {
        var scaledSize = newSize

        if image.size.width > image.size.height {
            let scaleFactor = image.size.width / image.size.height
            scaledSize.width = newSize.width
            scaledSize.height = newSize.height / scaleFactor
        } else {
            let scaleFactor = image.size.height / image.size.width
            scaledSize.height = newSize.height
            scaledSize.width = newSize.width / scaleFactor
        }

        let renderer = UIGraphicsImageRenderer(size: scaledSize)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: scaledSize))
        }
    }

    func signatureImage() -> UIImage? {
        return signatureImage(rotated: false, square: false)
    }

    func signatureImage(rotated: Bool) -> UIImage? {
        return signatureImage(rotated: rotated, square: false)
    }

    func signatureImage(rotated: Bool, square: Bool) -> UIImage? {
        guard hasSignature else { return nil }

        guard let snapshotImg = snapshot() else { return nil }
        erase()

        var signatureImg: UIImage

        if square {
            signatureImg = reduceImage(snapshotImg, to: CGSize(width: 400, height: 400))
        } else {
            signatureImg = snapshotImg
        }

        if rotated {
            signatureImg = rotateImage(signatureImg, clockwise: false)
        }

        return signatureImg
    }
}
